dataset/dataset_operations_common.h

```diff
@@ -229,4 +230,20 @@ void concatenate_out(const VariableConstView &var, const Dim dim,
   out_indices.assign(zip(out_begin, out_current));
 }
 
+/// Helper class for applying irreducible masks along dim.
+class Masker {
+public:
+  Masker(const DataArrayConstView &array, const Dim dim) {
+    const auto mask = irreducible_mask(array.masks(), dim);
+    if (mask)
+      m_masked = array.data() * ~mask;
+    m_data = m_masked ? m_masked : array.data();
+  }
+  auto data() const noexcept { return m_data; }
+
+private:
+  Variable m_masked;
+  VariableConstView m_data;
+};
```

Comment:
Why is a class necessary? Why can't this just be a function? I think in all cases you currently creat a Masker instance, you call `data()` on it almost immediately.

Reply:
`Masker` is the owner of `m_masked`. It has to be kept alive until the enclosing scope is left. This would not be necessary if our `VariableConstView` could *own* data, but as it is now it cannot.

Reply:
Ok I see why you can't do this now.
```
  Variable apply_masks_over_dim(const DataArrayConstView &array, const Dim dim) {
   Variable masked;
    const auto mask = irreducible_mask(array.masks(), dim);
    if (mask)
      masked = array.data() * ~mask;
    return masked ? masked : array.data();
  }
```

Reply:
Could you add a comment about this? Either here or as part of the clean up. I missed the purpose of the class as providing ownership when reviewing first time round.

Reply:
👍